src/state/swap/hooks.tsx

```diff
@@ -119,6 +120,18 @@ export function useDerivedSwapInfo(
     (isExactIn ? outputCurrency : inputCurrency) ?? undefined
   )
 
+  useEffect(() => {
```

Comment:
Can avoid a useEffect here and just do the conditional inline, instead of:

```tsx
useEffect(() => {
    if (trade.state === TradeState.INVALID) {
      setPreviouslyInvalid(true)
    } else if (trade.state !== TradeState.LOADING) {
      setPreviouslyInvalid(false)
    }
  }, [trade.state])
```

can do something like:

```tsx
const nextPreviouslyInvalid = (() => {
    if (trade.state === TradeState.INVALID) {
      return true
    } else if (trade.state !== TradeState.LOADING) {
      return false
    }
})()
if (typeof nextPreviouslyInvalid === 'boolean' && nextPreviouslyInvalid !== previouslyInvalid) {
  setPreviouslyInvalid(nextPreviouslyInvalid)
}
```

This will immediately bail from the current render and re-start, rather than rendering twice.

Reply:
Fixed! Out of curiosity is there a reason we want to avoid using useEffects?

Reply:
I'm also unclear about this. Feels more complicated to use a closure here?

Reply:
Closure isn't necessary it just imo is a way to group the value into one variable, certainly could be done any number of ways.

The reason for avoiding effect is just because it's not necessary, and effects have a cost because they run after the render, and then re-run another render. Whereas if you setState inside the render, you actual bail out of the render immediately and restart it, so you avoid finishing that render + you avoid a whole new render.

Great reading here https://react.dev/learn/you-might-not-need-an-effect

Reply:
i think this is clearer
```ts
let nextPreviouslyInvalid

if (trade.state === TradeState.INVALID) {
  nextPreviouslyInvalid = true
} else if (trade.state !== TradeState.LOADING) {
  nextPreviouslyInvalid = false
}
```